### DISCUSSION TITLE
Multi-threaded emulation logic

### DISCUSSION THREAD

**Discussion by lazerode** [86Box/86Box#3612] [Ideas]
_If we make 86Box's emulation logic multi-threaded, then will be able to emulate high-clocked CPUs on CPUs with poor IPSs, and add Pentium III and AMD Athlon emulation._

**The state of 86Box cycle-accurate emulation logic**: It is single threaded so if your CPU has poor instructions per seconds, high-clocked CPUs will be unbearly slow.

86Box 4.0 has 2 options: The old recompiler and the new recompiler. I am using the old recompiler for best stability and one day, I decided to try the 733MHz VIA Cyrix III (testing against my poor IPS Intel Core i5-1035G1) and see if it runs better than in 86Box 3.11 a bit. Well, it does (it can go further than 31% in 4.0 but not in 3.11 with my CPU), even with the old recompiler but that's still not enough.

**What to do**: Develop multi-threaded emulation that lets emulate fast CPUs in 86Box with CPUs that have poor instructions per second.

_**Note**: I never tried 86Box 4.0 with the new recompiler._

**Comment by Korkman** [comment]
The first thing that comes to mind is that if this was possible, modern physical CPUs would have boosted single-core performance through the roof by combining their many cores performance when a single threaded load appears (which, ironically, would solve this issue).

The x86 instruction set was never meant to be scalable horizontally. Engineers at Intel and AMD are squeezing a few more MHz out of the silicon with every release (mostly thanks to the advancements in structure size) to improve that single-thread performance. Instructions per cycle improvements have been made as well, mostly by introducing branch prediction (dating back to the first Pentiums), which CAN be parallelized, but is done in ultra-fast on-die cache memory.

And because branch prediction is already happening at the physical level, reinventing it in 86box is unlikely to have a positive impact on performance (btw. is branch prediction of Pentiums emulated in 86box? if so, kudos!).

My hopes are low there would be ways to distribute the emulation logic for a performance benefit. I would say hardware assisted emulation is more realistic (either by tapping into classic virtualization like qemu, vmware, etc. or FPGAs).

That being said, _recompilers_ are more about optimization than multi-threading. While yes, the recompilation step might benefit from many cores, the longer lasting execution benefits from clever tricks to make the emulated code run as native as possible in the confinements of a secure operating system. This is where some work is (was?) being done, hence the "New recompiler". I do hope the recompilers get revisited from time to time to see if improvements can be made. Cycle-accuracy is long gone when using recompilers, btw.

**Comment by Torinde** [comment]
- [Details on multi-threaded CPU emulation](https://github.com/joncampbell123/dosbox-x/issues/1184#issuecomment-514003096) possibilities at DOSBox-X
- https://github.com/86Box/86Box/discussions/3851
